File: go/use_title_case.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	// "golang.org/x/text/cases"
	// "golang.org/x/text/language"
)

// func main() {
// 	reader := bufio.NewReader(os.Stdin)
// 	fmt.Print("Enter text: ")
// 	text, _ := reader.ReadString('\n')

// 	titleCaser := cases.Title(language.English)
// 	title_cased_input := titleCaser.String(text)

// 	fmt.Println(title_cased_input)
// }

/*
This is an implementation without `cases` dependency.
You can also use the commented function above.
*/
func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter text: ")
	text, _ := reader.ReadString('\n')

	title_cased_input := use_title_case(text)
	fmt.Println(title_cased_input)
}

func contains(slice []string, item string) bool {
	for _, a := range slice {
		if a == item {
			return true
		}
	}
	return false
}
func use_title_case(input string) string {
	var title_cased_input string
	var words = strings.Split(input, " ")

	for i, word := range words {
		if contains(articles[:], word) || contains(prepositions[:], word) {
			title_cased_input += strings.ToLower(word)
		} else {
			uppercased_first_letter := strings.ToUpper(string(word[0]))
			var the_other_letters string

			for i := 1; i < len(word); i++ {
				the_other_letters += string(word[i])
			}

			title_cased_input += uppercased_first_letter + the_other_letters
		}

		if i != len(words)-1 {
			title_cased_input += " "
		}
	}

	return title_cased_input
}

var articles = [...]string{
	"a",
	"an",
	"the",
}
var prepositions = [...]string{
	"about",
	"above",
	"across",
	"after",
	"against",
	"along",
	"amid",
	"among",
	"around",
	"at",
	"before",
	"behind",
	"below",
	"beneath",
	"beside",
	"between",
	"beyond",
	"by",
	"concerning",
	"considering",
	"despite",
	"down",
	"during",
	"except",
	"for",
	"from",
	"in",
	"inside",
	"into",
	"like",
	"near",
	"of",
	"on",
	"onto",
	"out",
	"outside",
	"over",
	"past",
	"regarding",
	"round",
	"since",
	"through",
	"to",
	"toward",
	"under",
	"until",
	"up",
	"with",
	"within",
	"without",
}

File: rust/use_title_case.rs
fn main() {
    let input = "the world of technology";
    let output = use_title_case(input);

    println!("{}", output);
}

fn use_title_case(input: &str) -> String {
    let mut title_cased_string = String::new();
    let words: Vec<&str> = input.split(' ').collect();

    for (index, word) in words.iter().enumerate() {
        if word.is_empty() {
            continue;
        }

        if index != 0 && (PREPOSITIONS.contains(word) || ARTICLES.contains(word)) {
            title_cased_string.push_str(word);
        } else {
            let mut word_chars = word.chars();
            let first_letter = {
                let temp_first_letter = word_chars.next().unwrap();

                // Capitalize only when a word is alphabetic.
                if temp_first_letter.is_alphabetic() {
                    temp_first_letter.to_uppercase().next().unwrap()
                } else {
                    temp_first_letter
                }
            };

            let other_letters: String = word_chars.collect();
            title_cased_string.push_str(&format!("{}{}", first_letter, other_letters));
        }

        if index != words.len() - 1 {
            title_cased_string.push(' ');
        }
    }

    title_cased_string
}

const ARTICLES: [&str; 3] = ["a", "an", "the"];
const PREPOSITIONS: [&str; 50] = [
    "about",
    "above",
    "across",
    "after",
    "against",
    "along",
    "amid",
    "among",
    "around",
    "at",
    "before",
    "behind",
    "below",
    "beneath",
    "beside",
    "between",
    "beyond",
    "by",
    "concerning",
    "considering",
    "despite",
    "down",
    "during",
    "except",
    "for",
    "from",
    "in",
    "inside",
    "into",
    "like",
    "near",
    "of",
    "on",
    "onto",
    "out",
    "outside",
    "over",
    "past",
    "regarding",
    "round",
    "since",
    "through",
    "to",
    "toward",
    "under",
    "until",
    "up",
    "with",
    "within",
    "without",
];
